mezo-org/mezod
# EVM observability for the BTC Bridge

References: https://linear.app/thesis-co/issue/ENG-125/evm-observability-for-the-btc-bridge

### Introduction

As Bitcoin bridging occurs as part of the low-level consensus logic, we are unable to observe any Cosmos/EVM transactions. That means the only sign that bridging was completed on Mezo is the balance increase.
The information on bridged BTC assets is available within each block in a pseudo-transaction that is not stored within any database.

This PR enables storing such pseudo-transactions within indexer database so that they can be retrieved using commands like `eth_getTransactionByHash` just like ordinary transactions.

### Changes
- Extended the `TxResult` message with extra data field
- Added Storing the pseudo-transaction with `AssetsLocked` events in the indexer database
- Adjusted the following commands to work with pseudo-transactions: `eth_getTransactionByHash`,`eth_getTransactionReceipt`, `eth_getTransactionByBlockHashAndIndex`, `eth_getTransactionByBlockNumberAndIndex`, `eth_getBlockTransactionCountByHash`, `eth_getBlockTransactionCountByNumber`,`eth_getBlockByHash` and `eth_getBlockByNumber`, `debug_traceTransaction`, `debug_traceBlockByHash`, `debug_traceBlockByNumber`, `eth_getLogs`
- Added `AssetsBridge` precompile
- Added documentation 

### Testing

Local testing:
1. Set up a fresh local network of four nodes:
```
make localnet-bin-clean
make localnet-bin-init 
```

2. Enable indexer option in the config file of at least one node (e.g. modify this file `.localnet/node0/mezod/config/app.toml` by changing the option below from `false` to `true`)
```
...
# EnableIndexer enables the custom transaction indexer for the EVM (ethereum transactions).
enable-indexer = true
...
```

3. Launch the nodes in the network:
```
make localnet-bin-start // 4 times
```

4. Launch ethereum-sidecar. The sidecar should be able to provide the nodes in the network with information on assets bridged to Mezo. This can be done by connecting to some test network like Sepolia or by using some test sidecar, like the one we used before the real implementation of ethereum sidecar was added in this PR (#290). If you decide to use the test sidecar before the #290 PR merge you should replace these files with the old implementation:
```
ethereum/sidecar/cli/ethereum_sidecar.go
ethereum/sidecar/cli/flags.go
ethereum/sidecar/server.go
```

5. Set up accounts (called from `mezod/precompile/hardhat`):
```
npx hardhat run scripts/localhost-keys.ts | npx hardhat vars set MEZO_ACCOUNTS
```

6. List all the accounts:
```
npx hardhat --network localhost run scripts/accounts.ts
                       
0x0275e754739508629a2ECf6ae78fF78f6B80bcF5
0x031c4Fb2d908eAbAe6dC68d7A56485C1555EFBbb
0x9F6925C119355EC43200981026A06eA55623A16D
0x10B4951723C0F701a2f964ea690DF0F958E3597e
```

6. Add validator privileges to the first account:
```
npx hardhat validatorPool:addPrivilege --operators 0x0275e754739508629a2ECf6ae78fF78f6B80bcF5 --privilege-id 1 --signer 0x0275e754739508629a2ECf6ae78fF78f6B80bcF5
``` 

7. Set up the Blockscout explorer by cloning the repo:
```
git clone https://github.com/blockscout/blockscout.git
```

8. Go to `blockscout/docker-compose` and change these configuration options in `envs/common-blockscout.env`:
```
ETHEREUM_JSONRPC_HTTP_URL=http://0.0.0.0:8545
ETHEREUM_JSONRPC_WS_URL=ws://0.0.0.0:8546
NETWORK=Mezo
SUBNETWORK=Testnet
```

9. Build and launch the blockscout explorer:
```
docker compose up --build
```

10. Open a browser at http://localhost. You should see an explorer processing blocks and transactions.
     If you bridge some TBTC to Mezo you should see bridging transactions.

11. Verify the `AssetsBridge` smart contract so that the transactions are decoded:
Go the the http://localhost/api-docs and scroll down to `/smart-contracts/{address_hash}`, click to expand, then "try it out". Enter the precompile address (`0x7b7C000000000000000000000000000000000012`) and click execute. There should be a response with the status `200`.

If for some reason it does not work (the explorer does not recognize the `AssetsBridge` as a smart contract) then go to
any bridging transactions in the explorer and click the `to` field (it should be the `0x7b7C000000000000000000000000000000000012` address). After doing that the explorer should see the the address as a smart contract.

Execute `npx hardhat flatten contracts/AssetsBridgeCaller.sol > ~/Desktop/Bridge.sol` from `mezod/precompile/hardhat`.
In the blockscout menu select other > verify contract:
* Enter precompile address (`0x7b7C000000000000000000000000000000000012`)
* Select MIT license
* Verification method: Solidity (Single file)
* Compiler version: v0.8.24
* EVM Version: default
* Optimization enabled: true, runs: 200

Paste the contents of the flattened .sol created earlier into the contract code field. 
Hit Verify & publish.

---

### Author's checklist

- [x] Updated relevant unit and integration tests (will be done in the follow-up PR)

### Reviewer's checklist
